strawberryfields/io.py

```diff
@@ -0,0 +1,214 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Blackbird I/O
+=============
+
+This module contains functions for loading and saving Strawberry
+Fields :class:`~.Program` objects from/to Blackbird scripts.
+
+Summary
+-------
+
+.. autosummary::
+    to_blackbird
+    to_program
+    save
+    load
+
+Code details
+~~~~~~~~~~~~
+"""
+# pylint: disable=protected-access,too-many-nested-blocks
+import os
+import blackbird
+
+from . import ops
+from .program import Program
+
+
+def to_blackbird(prog, version="1.0"):
+    """Convert a Strawberry Fields Program to a Blackbird Program.
+
+    Args:
+        prog (Program): the Strawberry Fields program
+        version (str): Blackbird script version number
+
+    Returns:
+        blackbird.BlackbirdProgram:
+    """
+    bb = blackbird.BlackbirdProgram(name=prog.name, version=version)
+
+    if prog.backend is not None:
+        # set the target
+        bb._target["name"] = prog.backend
+
+    # fill in the quantum circuit
+    for cmd in prog.circuit:
+        op = {"kwargs": {}, "args": []}
+
+        op["op"] = cmd.op.__class__.__name__
+        op["modes"] = [i.ind for i in cmd.reg]
+
+        if "Measure" in op["op"]:
+            # special case to take into account 'select' keyword argument
+            if cmd.op.select is not None:
+                op["kwargs"]["select"] = cmd.op.select
+
+            if cmd.op.p:
+                # argument is quadrature phase
+                op["kwargs"]["phi"] = cmd.op.p[0].x
+
+        else:
+            if cmd.op.p is not None:
+                for a in cmd.op.p:
+                    # check if reg ref transform
+                    if isinstance(a.x, ops.RegRefTransform):
+                        if a.x.func_str is not None:
+                            # Note: not currently guaranteed to work,
+                            # as the RegRefTransform string cannot be checked
+                            # to determine if it is a valid function for serialization!
+                            #
+                            # Possible solutions:
+                            #
+                            #   * Use SymPy to represent functions, as
+                            #     SymPy provides methods for converting to
+                            #     Python functions as well as serialization
+                            #
+                            #   * Don't allow classical processing of measurements
```

Comment:
Seems like the lowest overhead way to do it. All post-processing can happen locally. Of course, this may mean more data transfer cost

Reply:
Perhaps any processing which does not change the size of a value in memory should be deferred to local device?

Reply:
Ah, but this is not post-processing, but mid-quantum simulation classical processing. So the remote backend will need the serialized function, it's not something that can be done locally.

I'm leaning towards a mixture of (a) SymPy approach for serializing any small function allowed by blackbird (i.e., basic arithmetic, sqrt, trig functions, exp, log), and (b) raising an error ("Function not supported on remote backend, please use a local backend") if the RegRefTransform includes anything more complicated.

Reply:
I think we should not allow inline processing for remote backends atm

Reply:
Removed from the PR.